api/osf_groups/serializers.py

```diff
@@ -0,0 +1,191 @@
+
+    def get_member_method(self, group, role):
+        methods = {
+            MANAGER: group.make_manager,
+            MEMBER: group.make_member,
+        }
+        return methods[role]
+
+    def get_group_role(self, validated_data, default_role):
+        role = validated_data.get('role', default_role)
+        if role not in GROUP_ROLES:
+            raise exceptions.ValidationError('{} is not a valid role; choose manager or member.'.format(role))
+        return role
+
+    class Meta:
+        type_ = 'group_members'
+
+    def get_absolute_url(self, obj):
+        return absolute_reverse(
+            'groups:group-member-detail',
+            kwargs={
+                'user_id': obj._id,
+                'group_id': self.context['request'].parser_context['kwargs']['group_id'],
+                'version': self.context['request'].parser_context['kwargs']['version'],
+            },
+        )
+
+
+class GroupMemberCreateSerializer(GroupMemberSerializer):
+    id = GroupCompoundIDField(source='_id', required=False, allow_null=True)
+    type = TypeField()
+    full_name = ser.CharField(required=False)
+    email = ser.EmailField(required=False, write_only=True)
+
+    def to_representation(self, instance, envelope='data'):
+        """
+        Use GroupMemberSerializer for the response, but GroupMemberCreateSerializer
+        for the request.  We only want full_name to be writable on create member (for unregistered members).
+        User serializer endpoints should be used to edit user's full_name.
+        """
+        return GroupMemberSerializer(instance=instance, context=self.context).data
+
+    def get_user_object(self, user_id, group):
+        if user_id:
+            user = OSFUser.load(user_id)
+            if not user:
+                raise exceptions.NotFound(detail='User with id {} not found.'.format(user_id))
+            if group.has_permission(user, 'member'):
+                raise exceptions.ValidationError(detail='User is already a member of this group.')
+            return user
+        return user_id
+
+    def create(self, validated_data):
+        group = self.context['group']
+        user = self.get_user_object(validated_data.get('_id', None), group)
+        auth = Auth(self.context['request'].user)
+        full_name = validated_data.get('full_name', None)
+        email = validated_data.get('email', None)
+        role = self.get_group_role(validated_data, MEMBER)
+
+        try:
+            if user:
+                self.get_member_method(group, role)(user, auth)
+            else:
+                if not full_name or not email:
+                    raise exceptions.ValidationError(detail='You must provide a full_name/email combination to add an unconfirmed member.')
+                else:
+                    user = group.add_unregistered_member(full_name, email, auth, role)
+        except ValueError as e:
+            # Making someone a member has the potential to violate the one-manager rule
```

Comment:
This comment isn't clear to me, it seems like making you mean taking away a manger role and making that a member role, but aren't all managers also members? Also I'm unsure where the ValueError is being thrown. I'd separate the registered and unregistered user code into two separate clauses, even if it's slightly longer.

Reply:
Sanity check: adding an unregistered member can't possibly violate the one manager rule, because you need one manager to even have a group to begin with.

Reply:
Yes, thank you, this comment doesn't describe why this is needed here.  (Background info: managers are also members, but they have the 'manager' permission while members have the 'member' permission.  If I'm a manager and I run OSFGroup.make_member(...) on myself, I am downgrading my manager perms. But that's not possible here).

However, both adding a member or an unregistered member can raise ValueErrors and I wanted to share this with both.   For example, if I'm trying to add an unregistered member that already exists, will raise a ValueError on the OSFGroup model.